src/pip/_internal/req/req_install.py

```diff
@@ -118,17 +117,22 @@ def __init__(
         self.req = req
         self.comes_from = comes_from
         self.constraint = constraint
-        if source_dir is None:
-            self.source_dir = None  # type: Optional[str]
-        else:
-            self.source_dir = os.path.normpath(os.path.abspath(source_dir))
         self.editable = editable
 
         if link is None and req and req.url:
             # PEP 508 URL requirement
             link = Link(req.url)
         self.link = self.original_link = link
         self.original_link_is_in_wheel_cache = False
+
+        self.source_dir = None  # type: Optional[str]
```

Comment:
A comment explaining what the non-None value means would nice. :)

Reply:
@pradyunsg I added a comment trying to explain what `source_dir` is and how it comes into existence.

BTW, has anyone ever considered unpacking lazily / on demand (e.g. when trying to obtain metadata). Instead of asserting everywhere in `InstallRequirement` that `source_dir` is set, create it if not yet present. I feel this could vastly simplify a lot of things.

Reply:
I've thought of that too; never came around to trying to implement it though. :)